astra/src/main/java/com/slack/astra/metadata/core/AstraMetadataStore.java

```diff
@@ -1,331 +1,708 @@
+   * @param mode the operation mode (overridden if either store is null)
+   * @param meterRegistry the metrics registry
+   */
   public AstraMetadataStore(
-      AsyncCuratorFramework curator,
-      AstraConfigs.ZookeeperConfig zkConfig,
-      CreateMode createMode,
-      boolean shouldCache,
-      ModelSerializer<T> modelSerializer,
-      String storeFolder,
+      ZookeeperMetadataStore<T> zkStore,
+      EtcdMetadataStore<T> etcdStore,
+      MetadataStoreMode mode,
       MeterRegistry meterRegistry) {
 
-    this.storeFolder = storeFolder;
-    this.zPath = ZPath.parseWithIds(String.format("%s/{name}", storeFolder));
-    this.zkConfig = zkConfig;
-    this.meterRegistry = meterRegistry;
-    String store = "/" + storeFolder.split("/")[1];
-
-    this.createCall = this.meterRegistry.counter(ASTRA_ZK_CREATE_CALL, "store", store);
-    this.deleteCall = this.meterRegistry.counter(ASTRA_ZK_DELETE_CALL, "store", store);
-    this.listCall = this.meterRegistry.counter(ASTRA_ZK_LIST_CALL, "store", store);
-    this.getCall = this.meterRegistry.counter(ASTRA_ZK_GET_CALL, "store", store);
-    this.hasCall = this.meterRegistry.counter(ASTRA_ZK_HAS_CALL, "store", store);
-    this.updateCall = this.meterRegistry.counter(ASTRA_ZK_UPDATE_CALL, "store", store);
-    this.addedListener = this.meterRegistry.counter(ASTRA_ZK_ADDED_LISTENER, "store", store);
-    this.removedListener = this.meterRegistry.counter(ASTRA_ZK_REMOVED_LISTENER, "store", store);
-    this.cacheInitializationHandlerFired =
-        this.meterRegistry.counter(ASTRA_ZK_CACHE_INIT_HANDLER_FIRED, "store", store);
-
-    ModelSpec<T> modelSpec =
-        ModelSpec.builder(modelSerializer)
-            .withPath(zPath)
-            .withCreateOptions(
-                Set.of(CreateOption.createParentsIfNeeded, CreateOption.createParentsAsContainers))
-            .withCreateMode(createMode)
-            .build();
-    modeledClient = ModeledFramework.wrap(curator, modelSpec);
-
-    if (shouldCache) {
-      cacheInitializedService =
-          Executors.newSingleThreadExecutor(
-              new ThreadFactoryBuilder().setNameFormat("cache-initialized-service-%d").build());
-      cachedModeledFramework = modeledClient.cached();
-      cachedModeledFramework.listenable().addListener(initializedListener, cacheInitializedService);
-      cachedModeledFramework.start();
+    this.zkStore = zkStore;
+    this.etcdStore = etcdStore;
+
+    // Override mode if one of the stores is null
+    if (zkStore == null && etcdStore != null) {
+      this.mode = MetadataStoreMode.ETCD_EXCLUSIVE;
+      LOG.info("ZK store is null, overriding mode to ETCD_EXCLUSIVE regardless of configured mode");
+    } else if (etcdStore == null && zkStore != null) {
+      this.mode = MetadataStoreMode.ZOOKEEPER_EXCLUSIVE;
+      LOG.info(
+          "Etcd store is null, overriding mode to ZOOKEEPER_EXCLUSIVE regardless of configured mode");
```

Comment:
How often would we hit these cases? Are these about to spew logs everytime a metadatastore is accessed?